rbd-mirror: separate ImageReplayer handling from Replayer

trociny opened the pull request:
Fixes: http://tracker.ceph.com/issues/18785
Signed-off-by: Mykola Golub <mgolub@mirantis.com>

dillaman commented on src/tools/rbd_mirror/CMakeLists.txt:
```diff
@@ -4,6 +4,7 @@ add_library(rbd_mirror_types STATIC
 set(rbd_mirror_internal
   ClusterWatcher.cc
   ImageDeleter.cc
+  ImageMapper.cc
```
I'd vote to skip class from this for this PR -- it's going to require a lot of refinement before it truly stubs out the necessary interactions

trociny replied:
The problem is that the Replayer needs to track in some way what images have already been acquired. It needs this to make a decision to release an image that was deleted in the peer pool. I expect it will be getting this info from ImageMapper in future, thus an extremely simplified version has been added.

I could just add a temporary ImageMapper class (or whatever it is called) to Replayer. Or add a method to InstanceReplayer to get e.g. a list of acquired images and use this instead, but I suppose this is not a way we a going to do this eventually so it would also be a temporary solution?

dillaman replied:
That's true, I guess -- fine to leave it. I should probably rebase #12364 and see if I can work around the recreated image race condition.

trociny commented:
@dillaman updated. It has not been tested well after update yet (running the tests), but I just wanted to know if you like this version -- I added handling of  repeated RPC message, removed "update" method and added  set_remote_images method to ImageReplayer , which is used now in acquire_image.

dillaman commented on src/tools/rbd_mirror/InstanceReplayer.cc:
```diff
@@ -0,0 +1,494 @@
+        auto local_image_id = image_replayer->get_local_image_id();
+        if (local_image_id.empty()) {
+          dout(20) << global_image_id << ": unknown local_image_id"
+                   << " (image not exist or primary), skipping delete" << dendl;
```
Nit: "... image does not exist ..."

dillaman commented on src/tools/rbd_mirror/InstanceReplayer.cc:
```diff
@@ -0,0 +1,494 @@
+        start_image_replayer(image_replayer);
+      }
+    });
+
```
I think we need to start and op here and finish it in the context above before each return point since we need to ensure zero percent possibility of the ```ImageDeleter``` calling back into a deleted instance replayer.

trociny commented on src/tools/rbd_mirror/InstanceReplayer.cc:
```diff
@@ -0,0 +1,494 @@
+  Context *ctx = new FunctionContext(
+    [this] (int r) {
+      cancel_image_state_check_task();
+      Mutex::Locker locker(m_lock);
```
Actually, the lock is not needed in wait_for_ops. I just wanted to make it similar to other xyz()/handle_xyz() functions.

Ok, I removed the lock in wait_for_ops (and create_async_context_callback, which is not needed after this).

trociny commented:
@dillaman updated

dillaman commented on src/tools/rbd_mirror/InstanceReplayer.cc:
```diff
@@ -0,0 +1,494 @@
+      dout(20) << "image deleter result: r=" << r << ", "
+               << "global_image_id=" << global_image_id << dendl;
+      if (r == -ESTALE || r == -ECANCELED) {
+        return;
```
```m_async_op_tracker.finish_op()```

dillaman commented on src/tools/rbd_mirror/InstanceReplayer.cc:
```diff
@@ -0,0 +1,494 @@
+      Mutex::Locker locker(m_lock);
+      auto it = m_image_replayers.find(global_image_id);
+      if (it == m_image_replayers.end()) {
+        return;
```
```m_async_op_tracker.finish_op()```

trociny commented:
@dillaman Thanks! I have just moved finish_op above returns.
